app/src/main/java/io/gnosis/safe/ui/transaction/TransactionsViewModel.kt

```diff
@@ -1,103 +1,122 @@
-        safe: Solidity.Address
-    ): List<TransactionView> {
-        return transactions.results.mapNotNull { transaction ->
-            when (transaction) {
-                is Transaction.Transfer -> TransactionView.Transfer(transaction, transaction.recipient == safe)
-                is Transaction.SettingsChange -> TransactionView.SettingsChange(transaction)
-                else -> null
+        currentSafeAddress = safe
+
+        val safeTxItems: Flow<PagingData<TransactionView>> = transactionsPagingRepository.getTransactionsStream(safe)
+            .map { pagingData ->
+                pagingData.map { transaction ->
+                    when (transaction) {
+                        is Transaction.Transfer -> TransactionView.Transfer(transaction, transaction.recipient == safe)
+                        is Transaction.SettingsChange -> TransactionView.SettingsChange(transaction)
+                        else -> TransactionView.Unknown
+                    }
+                }.filter { it !is TransactionView.Unknown }
             }
-        }
+            .map {
+                // insert headers
+                it.insertSeparators { before, after ->
+
+                    if (after == null) {
+                        // we're at the end of the list
+                        return@insertSeparators null  // no separator
+                    }
+
+                    if (before == null) {
+                        // we're at the beginning of the list
+                        return@insertSeparators when (after.transaction?.status) {
+                            TransactionStatus.Pending,
+                            TransactionStatus.AwaitingConfirmation,
+                            TransactionStatus.AwaitingExecution -> TransactionView.SectionHeader(title = R.string.tx_list_queue)
+
+                            TransactionStatus.Cancelled,
+                            TransactionStatus.Failed,
+                            TransactionStatus.Success -> TransactionView.SectionHeader(title = R.string.tx_list_history)
+
+                            else -> null  // no separator
+                        }
+                    }
+
+                    when (before.transaction?.status) {
+                        TransactionStatus.Pending,
+                        TransactionStatus.AwaitingConfirmation,
+                        TransactionStatus.AwaitingExecution -> when (after.transaction?.status) {
+                            TransactionStatus.Cancelled,
+                            TransactionStatus.Failed,
+                            TransactionStatus.Success -> TransactionView.SectionHeader(title = R.string.tx_list_history)
+                            else -> null  // no separator
+                        }
+                        else -> null  // no separator
+                    }
+                }
+            }
+            .cachedIn(viewModelScope)
+
+        currentSafeTxItems = safeTxItems
+        return safeTxItems
```

Comment:
Where are you consuming this return value? Maybe there is a way to avoid having an inner private `var` keeping track of it?

Reply:
it is used for caching - useful when navigating from and back to the fragment

Reply:
Is this consumed in the fragment? I can't seem to find the place where you use the value returned.

Reply:
it is consumed in the view model's load() method